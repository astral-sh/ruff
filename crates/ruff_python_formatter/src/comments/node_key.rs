use ruff_python_ast::AnyNodeRef;
use std::fmt::{Debug, Formatter};
use std::hash::{Hash, Hasher};

/// Used as key into the [`MultiMap`] storing the comments per node by [`Comments`].
///
/// Implements equality and hashing based on the address of the [`AnyNodeRef`] to get fast and cheap
/// hashing/equality comparison.
#[derive(Copy, Clone)]
pub(super) struct NodeRefEqualityKey<'a> {
    node: AnyNodeRef<'a>,
}

impl<'a> NodeRefEqualityKey<'a> {
    /// Creates a key for a node reference.
    pub(super) const fn from_ref(node: AnyNodeRef<'a>) -> Self {
        Self { node }
    }

    /// Returns the underlying node.
    pub(super) fn node(&self) -> AnyNodeRef {
        self.node
    }
}

impl Debug for NodeRefEqualityKey<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        self.node.fmt(f)
    }
}

impl PartialEq for NodeRefEqualityKey<'_> {
    fn eq(&self, other: &Self) -> bool {
        self.node.ptr_eq(other.node)
    }
}

impl Eq for NodeRefEqualityKey<'_> {}

impl Hash for NodeRefEqualityKey<'_> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.node.as_ptr().hash(state);
    }
}

impl<'a> From<AnyNodeRef<'a>> for NodeRefEqualityKey<'a> {
    fn from(value: AnyNodeRef<'a>) -> Self {
        NodeRefEqualityKey::from_ref(value)
    }
}

#[cfg(test)]
mod tests {
    use crate::comments::node_key::NodeRefEqualityKey;
    use ruff_python_ast::AnyNodeRef;
    use ruff_python_ast::StmtContinue;
    use ruff_text_size::TextRange;
    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};

    fn hash(key: NodeRefEqualityKey) -> u64 {
        let mut h = DefaultHasher::default();
        key.hash(&mut h);
        h.finish()
    }

    #[test]
    fn equality() {
        let continue_statement = StmtContinue {
            range: TextRange::default(),
        };

        let ref_a = NodeRefEqualityKey::from_ref(AnyNodeRef::from(&continue_statement));
        let ref_b = NodeRefEqualityKey::from_ref(AnyNodeRef::from(&continue_statement));

        assert_eq!(ref_a, ref_b);
        assert_eq!(hash(ref_a), hash(ref_b));
    }

    #[test]
    fn inequality() {
        let continue_statement = StmtContinue {
            range: TextRange::default(),
        };

        let boxed = Box::new(continue_statement.clone());

        let ref_a = NodeRefEqualityKey::from_ref(AnyNodeRef::from(&continue_statement));
        let ref_b = NodeRefEqualityKey::from_ref(AnyNodeRef::from(boxed.as_ref()));

        assert_ne!(ref_a, ref_b);
        assert_ne!(hash(ref_a), hash(ref_b));
    }
}
