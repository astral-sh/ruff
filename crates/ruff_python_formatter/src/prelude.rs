#[allow(unused_imports)]
pub(crate) use crate::{
    builders::PyFormatterExtensions, AsFormat, FormatNodeRule, FormattedIterExt as _, IntoFormat,
    PyFormatContext, PyFormatter,
};
#[allow(unused_imports)]
pub(crate) use ruff_formatter::prelude::*;
