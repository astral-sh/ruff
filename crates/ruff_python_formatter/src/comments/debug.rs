use std::fmt::{Debug, Formatter, Write};

use itertools::Itertools;

use ruff_formatter::SourceCode;
use ruff_text_size::Ranged;

use crate::comments::node_key::NodeRefEqualityKey;
use crate::comments::{CommentsMap, SourceComment};

/// Prints a debug representation of [`SourceComment`] that includes the comment's text
pub(crate) struct DebugComment<'a> {
    comment: &'a SourceComment,
    source_code: SourceCode<'a>,
}

impl<'a> DebugComment<'a> {
    pub(super) fn new(comment: &'a SourceComment, source_code: SourceCode<'a>) -> Self {
        Self {
            comment,
            source_code,
        }
    }
}

impl Debug for DebugComment<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut strut = f.debug_struct("SourceComment");

        strut
            .field("text", &self.comment.slice.text(self.source_code))
            .field("position", &self.comment.line_position)
            .field("formatted", &self.comment.formatted.get());

        strut.finish()
    }
}

/// Pretty-printed debug representation of [`Comments`].
pub(crate) struct DebugComments<'a> {
    comments: &'a CommentsMap<'a>,
    source_code: SourceCode<'a>,
}

impl<'a> DebugComments<'a> {
    pub(super) fn new(comments: &'a CommentsMap, source_code: SourceCode<'a>) -> Self {
        Self {
            comments,
            source_code,
        }
    }
}

impl Debug for DebugComments<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut map = f.debug_map();

        for node in self.comments.keys().sorted_by_key(|key| key.node().start()) {
            map.entry(
                &NodeKindWithSource {
                    key: *node,
                    source: self.source_code,
                },
                &DebugNodeComments {
                    comments: self.comments,
                    source_code: self.source_code,
                    key: *node,
                },
            );
        }

        map.finish()
    }
}

/// Prints the source code up to the first new line character. Truncates the text if it exceeds 40 characters.
struct NodeKindWithSource<'a> {
    key: NodeRefEqualityKey<'a>,
    source: SourceCode<'a>,
}

impl Debug for NodeKindWithSource<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        struct TruncatedSource<'a>(&'a str);

        impl Debug for TruncatedSource<'_> {
            fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
                f.write_char('`')?;
                let first_line = if let Some(line_end_pos) = self.0.find(['\n', '\r']) {
                    &self.0[..line_end_pos]
                } else {
                    self.0
                };

                if first_line.len() > 40 {
                    let (head, rest) = first_line.split_at(27);

                    f.write_str(head)?;
                    f.write_str("...")?;

                    // Take the last 10 characters
                    let tail = &rest[rest.len().saturating_sub(10)..];
                    f.write_str(tail)?;
                } else {
                    f.write_str(first_line)?;
                }

                if first_line.len() < self.0.len() {
                    f.write_str("\u{23ce}")?;
                }

                f.write_char('`')
            }
        }

        let kind = self.key.node().kind();
        let source = self.source.slice(self.key.node().range()).text(self.source);

        f.debug_struct("Node")
            .field("kind", &kind)
            .field("range", &self.key.node().range())
            .field("source", &TruncatedSource(source))
            .finish()
    }
}

struct DebugNodeComments<'a> {
    comments: &'a CommentsMap<'a>,
    source_code: SourceCode<'a>,
    key: NodeRefEqualityKey<'a>,
}

impl Debug for DebugNodeComments<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_map()
            .entry(
                &"leading",
                &DebugNodeCommentSlice {
                    node_comments: self.comments.leading(&self.key),
                    source_code: self.source_code,
                },
            )
            .entry(
                &"dangling",
                &DebugNodeCommentSlice {
                    node_comments: self.comments.dangling(&self.key),
                    source_code: self.source_code,
                },
            )
            .entry(
                &"trailing",
                &DebugNodeCommentSlice {
                    node_comments: self.comments.trailing(&self.key),
                    source_code: self.source_code,
                },
            )
            .finish()
    }
}

struct DebugNodeCommentSlice<'a> {
    node_comments: &'a [SourceComment],
    source_code: SourceCode<'a>,
}

impl Debug for DebugNodeCommentSlice<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut list = f.debug_list();

        for comment in self.node_comments {
            list.entry(&comment.debug(self.source_code));
        }

        list.finish()
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use ruff_formatter::SourceCode;
    use ruff_python_ast::AnyNode;
    use ruff_python_ast::{StmtBreak, StmtContinue};
    use ruff_python_trivia::CommentRanges;
    use ruff_text_size::{TextRange, TextSize};

    use crate::comments::map::MultiMap;
    use crate::comments::{CommentLinePosition, Comments, CommentsMap, SourceComment};

    #[test]
    fn debug() {
        let continue_statement = AnyNode::from(StmtContinue {
            range: TextRange::default(),
        });

        let break_statement = AnyNode::from(StmtBreak {
            range: TextRange::default(),
        });

        let source = r#"# leading comment
continue; # trailing
# break leading
break;
"#;

        let source_code = SourceCode::new(source);

        let mut comments_map: CommentsMap = MultiMap::new();

        comments_map.push_leading(
            continue_statement.as_ref().into(),
            SourceComment::new(
                source_code.slice(TextRange::at(TextSize::new(0), TextSize::new(17))),
                CommentLinePosition::OwnLine,
            ),
        );

        comments_map.push_trailing(
            continue_statement.as_ref().into(),
            SourceComment::new(
                source_code.slice(TextRange::at(TextSize::new(28), TextSize::new(10))),
                CommentLinePosition::EndOfLine,
            ),
        );

        comments_map.push_leading(
            break_statement.as_ref().into(),
            SourceComment::new(
                source_code.slice(TextRange::at(TextSize::new(39), TextSize::new(15))),
                CommentLinePosition::OwnLine,
            ),
        );

        let comment_ranges = CommentRanges::default();
        let comments = Comments::new(comments_map, &comment_ranges);

        assert_debug_snapshot!(comments.debug(source_code));
    }
}
