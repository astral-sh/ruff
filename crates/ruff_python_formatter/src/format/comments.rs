use crate::prelude::*;
use ruff_formatter::{write, Format};

use crate::cst::Attributed;
use crate::format::builders::literal;
use crate::trivia::TriviaKind;

#[derive(Debug)]
pub(crate) struct LeadingComments<'a, T> {
    item: &'a Attributed<T>,
}

impl<T> Format<ASTFormatContext<'_>> for LeadingComments<'_, T> {
    fn fmt(&self, f: &mut Formatter<ASTFormatContext>) -> FormatResult<()> {
        for trivia in &self.item.trivia {
            if trivia.relationship.is_leading() {
                match trivia.kind {
                    TriviaKind::EmptyLine => {
                        write!(f, [empty_line()])?;
                    }
                    TriviaKind::OwnLineComment(range) => {
                        write!(f, [literal(range, ContainsNewlines::No), hard_line_break()])?;
                    }
                    _ => {}
                }
            }
        }
        Ok(())
    }
}

#[inline]
pub(crate) const fn leading_comments<T>(item: &Attributed<T>) -> LeadingComments<'_, T> {
    LeadingComments { item }
}

#[derive(Debug)]
pub(crate) struct TrailingComments<'a, T> {
    item: &'a Attributed<T>,
}

impl<T> Format<ASTFormatContext<'_>> for TrailingComments<'_, T> {
    fn fmt(&self, f: &mut Formatter<ASTFormatContext>) -> FormatResult<()> {
        for trivia in &self.item.trivia {
            if trivia.relationship.is_trailing() {
                match trivia.kind {
                    TriviaKind::EmptyLine => {
                        write!(f, [empty_line()])?;
                    }
                    TriviaKind::OwnLineComment(range) => {
                        write!(f, [literal(range, ContainsNewlines::No), hard_line_break()])?;
                    }
                    _ => {}
                }
            }
        }
        Ok(())
    }
}

#[inline]
pub(crate) const fn trailing_comments<T>(item: &Attributed<T>) -> TrailingComments<'_, T> {
    TrailingComments { item }
}

#[derive(Debug)]
pub(crate) struct EndOfLineComments<'a, T> {
    item: &'a Attributed<T>,
}

impl<T> Format<ASTFormatContext<'_>> for EndOfLineComments<'_, T> {
    fn fmt(&self, f: &mut Formatter<ASTFormatContext>) -> FormatResult<()> {
        let mut first = true;
        for range in self
            .item
            .trivia
            .iter()
            .filter_map(|trivia| trivia.kind.end_of_line_comment())
        {
            if std::mem::take(&mut first) {
                write!(f, [line_suffix(&text("  "))])?;
            }
            write!(f, [line_suffix(&literal(range, ContainsNewlines::No))])?;
        }
        Ok(())
    }
}

#[inline]
pub(crate) const fn end_of_line_comments<T>(item: &Attributed<T>) -> EndOfLineComments<'_, T> {
    EndOfLineComments { item }
}

#[derive(Debug)]
pub(crate) struct DanglingComments<'a, T> {
    item: &'a Attributed<T>,
}

impl<T> Format<ASTFormatContext<'_>> for DanglingComments<'_, T> {
    fn fmt(&self, f: &mut Formatter<ASTFormatContext>) -> FormatResult<()> {
        for trivia in &self.item.trivia {
            if trivia.relationship.is_dangling() {
                if let TriviaKind::OwnLineComment(range) = trivia.kind {
                    write!(f, [hard_line_break()])?;
                    write!(f, [literal(range, ContainsNewlines::No)])?;
                    write!(f, [hard_line_break()])?;
                }
            }
        }
        Ok(())
    }
}

#[inline]
pub(crate) const fn dangling_comments<T>(item: &Attributed<T>) -> DanglingComments<'_, T> {
    DanglingComments { item }
}
