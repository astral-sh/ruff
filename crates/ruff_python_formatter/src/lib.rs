use thiserror::Error;
use tracing::Level;

use ruff_formatter::prelude::*;
use ruff_formatter::{format, FormatError, Formatted, PrintError, Printed, SourceCode};
use ruff_python_ast::node::AstNode;
use ruff_python_ast::Mod;
use ruff_python_index::tokens_and_ranges;
use ruff_python_parser::lexer::LexicalError;
use ruff_python_parser::{parse_ok_tokens, Mode, ParseError};
use ruff_python_trivia::CommentRanges;
use ruff_source_file::Locator;

use crate::comments::{
    dangling_comments, leading_comments, trailing_comments, Comments, SourceComment,
};
pub use crate::context::PyFormatContext;
pub use crate::options::{MagicTrailingComma, PreviewMode, PyFormatOptions, QuoteStyle};
use crate::verbatim::suppressed_node;

pub(crate) mod builders;
pub mod cli;
mod comments;
pub(crate) mod context;
pub(crate) mod expression;
mod generated;
pub(crate) mod module;
mod options;
pub(crate) mod other;
pub(crate) mod pattern;
mod prelude;
pub(crate) mod statement;
pub(crate) mod type_param;
mod verbatim;

include!("../../ruff_formatter/shared_traits.rs");

/// 'ast is the lifetime of the source code (input), 'buf is the lifetime of the buffer (output)
pub(crate) type PyFormatter<'ast, 'buf> = Formatter<'buf, PyFormatContext<'ast>>;

/// Rule for formatting a Python [`AstNode`].
pub(crate) trait FormatNodeRule<N>
where
    N: AstNode,
{
    fn fmt(&self, node: &N, f: &mut PyFormatter) -> FormatResult<()> {
        let comments = f.context().comments().clone();

        let node_comments = comments.leading_dangling_trailing(node.as_any_node_ref());

        if self.is_suppressed(node_comments.trailing, f.context()) {
            suppressed_node(node.as_any_node_ref()).fmt(f)
        } else {
            leading_comments(node_comments.leading).fmt(f)?;

            let is_source_map_enabled = f.options().source_map_generation().is_enabled();

            if is_source_map_enabled {
                source_position(node.start()).fmt(f)?;
            }

            self.fmt_fields(node, f)?;
            self.fmt_dangling_comments(node_comments.dangling, f)?;

            if is_source_map_enabled {
                source_position(node.end()).fmt(f)?;
            }

            trailing_comments(node_comments.trailing).fmt(f)
        }
    }

    /// Formats the node's fields.
    fn fmt_fields(&self, item: &N, f: &mut PyFormatter) -> FormatResult<()>;

    /// Formats the [dangling comments](comments#dangling-comments) of the node.
    ///
    /// You should override this method if the node handled by this rule can have dangling comments because the
    /// default implementation formats the dangling comments at the end of the node, which isn't ideal but ensures that
    /// no comments are dropped.
    ///
    /// A node can have dangling comments if all its children are tokens or if all node children are optional.
    fn fmt_dangling_comments(
        &self,
        dangling_node_comments: &[SourceComment],
        f: &mut PyFormatter,
    ) -> FormatResult<()> {
        debug_assert!(
            dangling_node_comments.is_empty(),
            "The node has dangling comments that need to be formatted manually. Add the special dangling comments handling to `fmt_fields` and override `fmt_dangling_comments` with an empty implementation that returns `Ok(())`."
        );

        dangling_comments(dangling_node_comments).fmt(f)
    }

    fn is_suppressed(
        &self,
        _trailing_comments: &[SourceComment],
        _context: &PyFormatContext,
    ) -> bool {
        false
    }
}

#[derive(Error, Debug)]
pub enum FormatModuleError {
    #[error("source contains syntax errors: {0:?}")]
    LexError(LexicalError),
    #[error("source contains syntax errors: {0:?}")]
    ParseError(ParseError),
    #[error(transparent)]
    FormatError(#[from] FormatError),
    #[error(transparent)]
    PrintError(#[from] PrintError),
}

impl From<LexicalError> for FormatModuleError {
    fn from(value: LexicalError) -> Self {
        Self::LexError(value)
    }
}

impl From<ParseError> for FormatModuleError {
    fn from(value: ParseError) -> Self {
        Self::ParseError(value)
    }
}

#[tracing::instrument(name = "format", level = Level::TRACE, skip_all)]
pub fn format_module_source(
    source: &str,
    options: PyFormatOptions,
) -> Result<Printed, FormatModuleError> {
    let (tokens, comment_ranges) = tokens_and_ranges(source)?;
    let module = parse_ok_tokens(tokens, source, Mode::Module, "<filename>")?;
    let formatted = format_module_ast(&module, &comment_ranges, source, options)?;
    Ok(formatted.print()?)
}

pub fn format_module_ast<'a>(
    module: &'a Mod,
    comment_ranges: &'a CommentRanges,
    source: &'a str,
    options: PyFormatOptions,
) -> FormatResult<Formatted<PyFormatContext<'a>>> {
    let source_code = SourceCode::new(source);
    let comments = Comments::from_ast(module, source_code, comment_ranges);
    let locator = Locator::new(source);

    let formatted = format!(
        PyFormatContext::new(options, locator.contents(), comments),
        [module.format()]
    )?;
    formatted
        .context()
        .comments()
        .assert_all_formatted(source_code);
    Ok(formatted)
}

/// Public function for generating a printable string of the debug comments.
pub fn pretty_comments(module: &Mod, comment_ranges: &CommentRanges, source: &str) -> String {
    let source_code = SourceCode::new(source);
    let comments = Comments::from_ast(module, source_code, comment_ranges);

    std::format!("{comments:#?}", comments = comments.debug(source_code))
}

#[cfg(test)]
mod tests {
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;

    use ruff_python_index::tokens_and_ranges;

    use ruff_python_parser::{parse_ok_tokens, Mode};

    use crate::{format_module_ast, format_module_source, PyFormatOptions};

    /// Very basic test intentionally kept very similar to the CLI
    #[test]
    fn basic() -> Result<()> {
        let input = r#"
# preceding
if    True:
    pass
# trailing
"#;
        let expected = r#"# preceding
if True:
    pass
# trailing
"#;
        let actual = format_module_source(input, PyFormatOptions::default())?
            .as_code()
            .to_string();
        assert_eq!(expected, actual);
        Ok(())
    }

    /// Use this test to debug the formatting of some snipped
    #[ignore]
    #[test]
    fn quick_test() {
        let source = r#"
def main() -> None:
    if True:
        some_very_long_variable_name_abcdefghijk = Foo()
        some_very_long_variable_name_abcdefghijk = some_very_long_variable_name_abcdefghijk[
            some_very_long_variable_name_abcdefghijk.some_very_long_attribute_name
            == "This is a very long string abcdefghijk"
        ]

"#;
        let (tokens, comment_ranges) = tokens_and_ranges(source).unwrap();

        // Parse the AST.
        let source_path = "code_inline.py";
        let module = parse_ok_tokens(tokens, source, Mode::Module, source_path).unwrap();
        let options = PyFormatOptions::from_extension(Path::new(source_path));
        let formatted = format_module_ast(&module, &comment_ranges, source, options).unwrap();

        // Uncomment the `dbg` to print the IR.
        // Use `dbg_write!(f, []) instead of `write!(f, [])` in your formatting code to print some IR
        // inside of a `Format` implementation
        // use ruff_formatter::FormatContext;
        // dbg!(formatted
        //     .document()
        //     .display(formatted.context().source_code()));
        //
        // dbg!(formatted
        //     .context()
        //     .comments()
        //     .debug(formatted.context().source_code()));

        let printed = formatted.print().unwrap();

        assert_eq!(
            printed.as_code(),
            r#"for converter in connection.ops.get_db_converters(
    expression
) + expression.get_db_converters(connection):
    ...
"#
        );
    }

    #[test]
    fn string_processing() {
        use crate::prelude::*;
        use ruff_formatter::{format, format_args, write};

        struct FormatString<'a>(&'a str);

        impl Format<SimpleFormatContext> for FormatString<'_> {
            fn fmt(&self, f: &mut Formatter<SimpleFormatContext>) -> FormatResult<()> {
                let format_str = format_with(|f| {
                    write!(f, [token("\"")])?;

                    let mut words = self.0.split_whitespace().peekable();
                    let mut fill = f.fill();

                    let separator = format_with(|f| {
                        group(&format_args![
                            if_group_breaks(&token("\"")),
                            soft_line_break_or_space(),
                            if_group_breaks(&token("\" "))
                        ])
                        .fmt(f)
                    });

                    while let Some(word) = words.next() {
                        let is_last = words.peek().is_none();
                        let format_word = format_with(|f| {
                            write!(f, [text(word, None)])?;

                            if is_last {
                                write!(f, [token("\"")])?;
                            }

                            Ok(())
                        });

                        fill.entry(&separator, &format_word);
                    }

                    fill.finish()
                });

                write!(
                    f,
                    [group(&format_args![
                        if_group_breaks(&token("(")),
                        soft_block_indent(&format_str),
                        if_group_breaks(&token(")"))
                    ])]
                )
            }
        }

        // 77 after g group (leading quote)
        let fits =
            r#"aaaaaaaaaa bbbbbbbbbb cccccccccc dddddddddd eeeeeeeeee ffffffffff gggggggggg h"#;
        let breaks =
            r#"aaaaaaaaaa bbbbbbbbbb cccccccccc dddddddddd eeeeeeeeee ffffffffff gggggggggg hh"#;

        let output = format!(
            SimpleFormatContext::default(),
            [FormatString(fits), hard_line_break(), FormatString(breaks)]
        )
        .expect("Formatting to succeed");

        assert_snapshot!(output.print().expect("Printing to succeed").as_code());
    }
}
