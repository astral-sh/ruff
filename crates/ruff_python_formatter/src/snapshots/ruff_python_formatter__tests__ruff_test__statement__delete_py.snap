---
source: crates/ruff_python_formatter/src/lib.rs
expression: snapshot
---
## Input
```py
x = 1
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa = 1
b, c, d = (2, 3, 4)

# Delete something
del x  # Deleted something
# Done deleting

# Delete something
del (
    # Deleting something
    x  # Deleted something
    # Finishing deletes
) # Completed
# Done deleting

# Delete something
del (
    # Deleting something
    x  # Deleted something
    # Finishing deletes
) # Completed
# Done deleting

# Delete something
del (
    # Deleting something
    x,  # Deleted something
    # Finishing deletes
) # Completed
# Done deleting

# Delete something
del (
    # Deleting something
    x  # Deleted something
    # Finishing deletes

    # Dangling comment
) # Completed
# Done deleting

# Delete something
del x, aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa, b, c, d  # Delete these
# Ready to delete

# Delete something
del (
    x,
    # Deleting this
    aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa,
    b,
    c,
    d
    # Deleted
) # Completed
# Done
```



## Output
```py
x = 1
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa = 1
b, c, d = (2, 3, 4)

# Delete something
del x  # Deleted something
# Done deleting

# Delete something
del (
    # Deleting something
    x  # Deleted something
    # Finishing deletes
    # Completed
)
# Done deleting

# Delete something
del (
    # Deleting something
    x  # Deleted something
    # Finishing deletes
    # Completed
)
# Done deleting

# Delete something
del (
    # Deleting something
    x,  # Deleted something
    # Finishing deletes
)  # Completed
# Done deleting

# Delete something
del (
    # Deleting something
    x  # Deleted something
    # Finishing deletes
    # Dangling comment
    # Completed
)
# Done deleting

# Delete something
del (
    x,
    aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa,
    b,
    c,
    d,
)  # Delete these
# Ready to delete

# Delete something
del (
    x,
    # Deleting this
    aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa,
    b,
    c,
    d
    # Deleted
)  # Completed
# Done
```


